aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -0,0 +1,91 @@
+#include "ATen/ATen.h"
+#include "ATen/Dispatch.h"
+#include "ATen/ExpandUtils.h"
+#include "ATen/NativeFunctions.h"
+#include "ATen/WrapDimUtils.h"
+#include "ATen/cpu/cpuinfo/include/cpuinfo.h"
+#include "cpu/ReduceOpsKernel.h"
+
+#include <algorithm>
+#include <functional>
+#include <numeric>
+#include <vector>
+
+#include <map>
+
+namespace at {
+namespace native {
+
+sumallF sumallDispatch;
+sumallF *sumall = &sumallDispatch;
+void sumallDispatch(Tensor &result, const Tensor &self) {
+  cpuinfo_initialize();
+  sumall = &sumallNONE;
+  if (cpuinfo_has_x86_avx()) {
+    sumall = &sumallAVX;
+  }
+  if (cpuinfo_has_x86_avx2()) {
+    sumall = &sumallAVX2;
+  }
+  return (*sumall)(result, self);
+}
+
+reducesumF reducesumDispatch;
+reducesumF *reducesum = &reducesumDispatch;
+void reducesumDispatch(Tensor &result, const Tensor &self, size_t dim) {
+  cpuinfo_initialize();
+  reducesum = &reducesumNONE;
+  if (cpuinfo_has_x86_avx()) {
+    reducesum = &reducesumAVX;
+  }
+  if (cpuinfo_has_x86_avx2()) {
+    reducesum = &reducesumAVX2;
+  }
+  return (*reducesum)(result, self, dim);
+}
```

Comment:
I played a bit with ways of listing the possible implementations and came up with [something like this](https://godbolt.org/g/rqKjtL). The implementation is quite short, and everything that's needed to implement the dispatch of `sumAll` is this:
```cpp
using sum_type = void(const float *a, float *b);
extern sum_type sumAllAVX;
extern sum_type sumAllNone;
struct ImplSpec<sum_type> sumAllImpls[] = {
  { CPUCapability::AVX,  &sumAllAVX   },
  { CPUCapability::NONE, &sumAllNone  },
};
FnPtr<sum_type> sumAll = &DispatchStub<sum_type>::init<sumAllImpls, &sumAll>;
```
It will automatically generate a stub that will check CPU capability when called for the first time, and later replace the function pointer with the best concrete implementation.

Reply:
I'll also look into being able to get rid of manually specifying any of these suffixes. Ideally you'd just implement it using the macro and the only while that actually cares about the capabilities is vec256, unless you really want to care.